add LACP and IPv6 ACLs

cjdoucette opened the pull request:
The first patch adds LACP, and takes steps to ensure its requirements are almost always met (but they are not always guaranteed). There is also an important TODO left in LACP, which is to not report dropped packets when LACP is configuring, and to warn the user when LACP configuration is taking too long. This turned out to be harder than I expected and I need to figure out a good way of doing it.

The second patch cleans up some configuration inconsistencies that I noticed in preparation for the third patch.

The third patch adds IPv6 ACLs to GK, GT, and LLS, so that ND packets (and eventually BGP packets) can be recognized quickly. This patch also reverses some changes that were previously made to recognize ND packets in those blocks.

AltraMayor commented on lib/net.c:
```diff
@@ -1219,6 +1217,13 @@ gatekeeper_init_network(struct net_config *net_conf)
 		goto out;
 	}
 
+	if (GATEKEEPER_MAX_PKT_BURST < 2 * GATEKEEPER_MAX_PORTS) {
```
Shouldn't this test only be performed if LACP is enabled?

AltraMayor commented on lls/main.c:
```diff
@@ -224,6 +253,34 @@ lls_scan(__attribute__((unused)) struct rte_timer *timer, void *arg)
+			xmit_arp_req(&net_conf->front,
+				(const uint8_t *)&net_conf->front.ip4_addr,
+				NULL, lls_conf->tx_queue_front);
+		else
```
It would be nice to have an `else if (likely(iface_nd_enabled(...)))` followed by an `else rte_panic(...)` here and in analog code below just to quickly catch hard bugs when they show up.

AltraMayor commented on lls/main.c:
```diff
@@ -337,6 +394,15 @@ lls_proc(void *arg)
+			 * the LLS block receives a lot of requests but
+			 * we are unable to answer them -- i.e. the
+			 * number of requests > 0 for a sustained
+			 * period but we never invoke the TX burst.
```
You could modify `process_pkts()` above to return how many packets were transmitted to reset the LACP timer since that it can now skip a cycle. Another opportunity to reset the LACP timer is during a scan of the caches if a packet is transmited.

cjdoucette replied:
I modified `process_pkts()` to do this, but didn't touch the timer during a the cache scan since it involved changing a lot of function prototypes and is a rare opportunity to skip a cycle anyway.

AltraMayor replied:
Ok.

AltraMayor commented on lua/gatekeeper.lua:
```diff
@@ -99,6 +99,11 @@ local ffi = require("ffi")
 
+struct constants {
+	static const uint32_t BONDING_MODE_ROUND_ROBIN = 0;
+	static const uint32_t BONDING_MODE_8023AD = 4;
```
1. Why didn't you include the other modes?

2. Check the examples in `lua/policylib.lua` on how to bring C constants into Lua. It's cleaner than the way you're doing now.

cjdoucette replied:
1. I didn't include the other modes because I haven't seen a need for them in Gatekeeper, and I didn't want to add more risk in adding them without purpose. Looking at the documentation, it seems only the LACP mode requires further application restrictions -- the rest should be naturally supported. Should I add the others?

2. Thanks, will do.

AltraMayor replied:
I think it's a good idea to add the other modes since they only require the constants in Lua.

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -242,6 +246,20 @@ extern uint8_t rss_key_be[RTE_DIM(default_rss_key)];
+	/* When @iface is the back, need to make sure it's enabled. */
+	if (iface == &net->back)
+		return net->back_iface_enabled &&
+			iface->num_ports > 1 &&
```
The Internet Exchange PPT-SP requires LACP even on a single interface.

cjdoucette replied:
Ah, I see. Just to confirm: this means that even when there is a single port, we should create a bonded port if LACP is being used (in contrast to the current initialization, which does not do round-robin bonding on the port at all if there's only one).

AltraMayor replied:
Yes, you got it right, if LACP is required, we create the bonded port even for a single interface.

cjdoucette commented:
All the suggestions so far have been addressed in the latest rebase.

AltraMayor commented on lls/main.c:
```diff
@@ -224,13 +253,46 @@ lls_scan(__attribute__((unused)) struct rte_timer *timer, void *arg)
+	}
+}
+
+static uint16_t
```
Use `int` here. There's no reason to use another type here.

AltraMayor commented on lls/main.c:
```diff
@@ -319,13 +385,25 @@ lls_proc(void *arg)
 				lls_conf->tx_queue_back);
 
+		if (num_tx > 0 && (lacp_enabled(net_conf, &net_conf->front) ||
+				lacp_enabled(net_conf, &net_conf->back))) {
```
There's a bug here. If the first `process_pkts()` send out a packet and the second `process_pkts()` didn't send out a packet, the code will reset the timer for both interfaces.

We should try to skip those ARP and ND gratuitous as much as we can without compromising the performance of Gatekeeper. I think the straightforward solution is to have a LACP timer per interface.

AltraMayor commented on lls/main.c:
```diff
@@ -319,13 +385,25 @@ lls_proc(void *arg)
 
+		if (num_tx > 0 && (lacp_enabled(net_conf, &net_conf->front) ||
+				lacp_enabled(net_conf, &net_conf->back))) {
+			int ret = rte_timer_reset(&lls_conf->lacp_timer,
```
The call of `rte_timer_reset()` on `lls_conf->lacp_timer` is quite special, i.e., many specific parameters. It'd be better to encapsulate it into an inline function and replace all current calls with the new inline function.

AltraMayor commented on lls/main.c:
```diff
@@ -224,13 +253,46 @@ lls_scan(__attribute__((unused)) struct rte_timer *timer, void *arg)
+			xmit_arp_req(&net_conf->front,
+				(const uint8_t *)&net_conf->front.ip4_addr,
+				NULL, lls_conf->tx_queue_front);
+		else if (likely(iface_nd_enabled(net_conf, &net_conf->front)))
```
This is the first occasion we want to test IPv6 (i.e. ND) before IPv4 (i.e. ARP): ARP broadcast gratuitous packets whereas ND multicast gratuitous packets. That is, ND gratuitous packets put less load on the network fabric. Could invert the test here and below?

AltraMayor replied:
Please also add a comment explaining the reasoning.

AltraMayor commented on gk/main.c:
```diff
@@ -707,15 +707,15 @@ gk_stage1(void *arg)
 			RTE_LOG(ERR, GATEKEEPER, "gk: cannot assign an RX queue for the front interface for lcore %u\n",
 				lcore);
-			return -1;
+			goto cleanup;
```
Shouldn't the `return -1;` before the `for` statement become a `goto cleanup`? Notice that I cannot comment there since the code wasn't changed there.

mengxiang0811 commented on gt/main.c:
```diff
@@ -597,6 +585,33 @@ gt_proc(void *arg)
+					__func__);
+				for (i = 0; i < acl.num; i++)
+					rte_pktmbuf_free(acl.mbufs[i]);
+				break;
```
Just to confirm, do you mean to quit the gt instance here?

cjdoucette replied:
Yes, since if the ACL classification is wrong then it's a logical error, not a problem at runtime. Same with the GK block below.

mengxiang0811 commented on gk/main.c:
```diff
@@ -634,6 +711,33 @@ gk_proc(void *arg)
+					__func__);
+				for (i = 0; i < acl.num; i++)
+					drop_packet(acl.mbufs[i]);
+				break;
```
Just to confirm, do you mean to quit the gk instance here?

mengxiang0811 commented on gk/main.c:
```diff
@@ -634,6 +711,33 @@ gk_proc(void *arg)
+
+		if (ipv6_if_configured(&gk_conf->net->front) && acl.num > 0) {
+			ret = rte_acl_classify(
+			    gk_conf->ipv6_acls[rte_lcore_to_socket_id(lcore)],
```
It seems that here is not a tab space.

cjdoucette replied:
Fixed.

mengxiang0811 commented on gt/main.c:
```diff
@@ -505,30 +509,14 @@ gt_proc(void *arg)
-						    &gt_conf->net->front) == -1)
-						rte_pktmbuf_free(m);
+				if (ipv6_if_configured(&gt_conf->net->front) &&
+				    pkt_info.outer_ip_ver == ETHER_TYPE_IPv6) {
```
Not a tab space here.

cjdoucette replied:
Thanks! Fixed.

cjdoucette commented:
The above suggestions are now integrated.

cjdoucette commented on lua/net.lua:
```diff
@@ -25,6 +27,7 @@ return function (gatekeeper_server)
 	local front_iface = gatekeeper.c.get_if_front(net_conf)
 	front_iface.arp_cache_timeout_sec = front_arp_cache_timeout_sec
 	front_iface.nd_cache_timeout_sec = front_nd_cache_timeout_sec
+	front_iface.bonding_mode = front_bonding_mode
```
I think we should try to keep all of the code that the user can change in one place, so that they don't need to change anything below the `Code below this point should not need to be changed.` warning.

cjdoucette commented:
And Qiaobin's suggestions have been added, too.

AltraMayor commented on lls/main.c:
```diff
@@ -224,13 +254,50 @@ lls_scan(__attribute__((unused)) struct rte_timer *timer, void *arg)
 
+/* Send gratuitous ARP or ND announcement to fulfill LACP TX requirement. */
 static void
+lls_lacp_announce(__attribute__((unused)) struct rte_timer *timer, void *arg)
```
@cjdoucette, could you read DPDK's bonding driver to see if we can call the transmitting function with no packets? I'm concerned that this uncommon behavior, namely, sending gratuitous packet often, can lead an IX to complain about Gatekeeper.

cjdoucette replied:
I had assumed that the TX function would fail if given if zero packets, but thanks for the idea to check that assumption! It actually works fine with transmitting zero packets, and both reduces the number of unnecessary packets and simplifies the code. I'll make the changes now.

AltraMayor commented on lls/main.c:
```diff
@@ -21,14 +21,39 @@
+ *    TX burst requirement on a timer that runs slightly more frequently
+ *    than every 100ms, defined below.
+ */
+#define LLS_LACP_ANNOUNCE_INTERVAL_MS 95
```
1ms is a lot for Gatekeeper. I think it's safe to use 99ms here.

cjdoucette replied:
OK, changed it.

AltraMayor commented on lls/main.c:
```diff
@@ -337,6 +409,15 @@ lls_proc(void *arg)
 			 * timer is inefficient. See the timer application.
+			 *
+			 * Also send gratuitous ARP/ND messages to fulfill
+			 * the LACP requirement.
```
The comment above became outdated, didn't it?

cjdoucette replied:
Yes, fixed that now. Thanks!

AltraMayor commented on gk/main.c:
```diff
@@ -654,11 +654,11 @@ cleanup_gk(struct gk_config *gk_conf)
 
 	for (i = 0; i < gk_conf->num_lcores; i++) {
-		if (gk_conf->instances[i].ip_flow_hash_table)
+		if (gk_conf->instances[i].ip_flow_hash_table == NULL)
```
Shouldn't it be `!=` instead of `==`?

AltraMayor commented on gk/main.c:
```diff
@@ -654,11 +654,11 @@ cleanup_gk(struct gk_config *gk_conf)
 				ip_flow_hash_table);
 
-		if (gk_conf->instances[i].ip_flow_entry_table)
+		if (gk_conf->instances[i].ip_flow_entry_table == NULL)
```
Shouldn't it be `!=` instead of `==`?

AltraMayor commented on gk/main.c:
```diff
@@ -634,6 +711,34 @@ gk_proc(void *arg)
         	}
+
+		if (ipv6_if_configured(&gk_conf->net->front) && acl.num > 0) {
+			ret = rte_acl_classify(
```
Why didn't you use ` rte_acl_classify_alg` here?

AltraMayor replied:
Ok, `rte_acl_classify()` uses the algorithm set in the ACL context, and I found your XXX comment pushing the use of `rte_acl_set_ctx_classify()` to later. Is it hard to find the supported hardware and set the right algorithm?

AltraMayor replied:
There's `rte_acl_init()` in `dependencies/dpdk/lib/librte_acl/rte_acl.c` that detects the best algorithm available, but for some reason `rte_acl_init()` is `static` and never called. Isn't asking in the mailing list the why worth?

AltraMayor replied:
Ok, we don't need to do anything, DPDK will pick the most efficient algorithm available automatically.

cjdoucette replied:
I see the `__attribute__((constructor))` now on `rte_acl_init()`, so I agree that the ACL library will take care of it -- we can remove the `XXX` about changing the classification.

AltraMayor commented on lib/mailbox.c:
```diff
@@ -27,6 +27,11 @@
+/*
+ * XXX Add ability to bulk enqueue/dequeue from a mailbox to speed
+ * up performance of transfer of IPv6 packets between blocks.
+ */
```
Isn't it easy and worth to add?

cjdoucette replied:
This turned out to be unnecessary, as a *single* request (using one allocation from the LLS mailbox) can be used to hold an array of ND packets, achieving the same effect. The `submit_nd()` function changes to accommodate multiple ND packets in the latest patch, but the mailbox didn't need to change.

cjdoucette commented on gk/main.c:
```diff
@@ -654,6 +759,13 @@ cleanup_gk(struct gk_config *gk_conf)
+		if (ipv6_if_configured(&gk_conf->net->front) &&
+				gk_conf->ipv6_acls[socket] != NULL) {
+			rte_acl_free(gk_conf->ipv6_acls[socket]);
+			gk_conf->ipv6_acls[socket] = NULL;
```
Yes, we could have IPv6 ACLs for each interface instead of letting each block have their own. This would require a small adjustment though -- I'll describe it so that we are on the same page.

Right now, ACLs are allocated on a per-socket basis -- for example, each GK lcore doesn't need its own ACL, so all the lcores on the same socket share an ACL.

A problem arises when another block is trying to use the IPv6 ACL on the same socket that the GK lcores are using. The rules that this new block wants to add to the ACL may conflict with the rules installed by GK.

DPDK has a solution for this, which is that when classifying packets, it can return matches in different categories. So even if multiple blocks are using the same ACL for different rules, the classification algorithm can return one match for the GK-related rules and a match for the new block's rules.

I'll take this suggestion and the ones above to refactor the ACL stuff away from the blocks.

AltraMayor replied:
Having ACLs allocated per socket is a good idea since this enables each socket to quickly access the ACLs. This is still a good idea even accounting for the work to keep these ACLs identical.

You should think of the IPv6 filter in the same way we think of filter in the hardware: if there are conflicts or the filter is wrongly defined, it's a bug.

AltraMayor replied:
Having ACLs allocated per socket is a good idea since this enables each socket to quickly access the ACLs. This is still a good idea even accounting for the work to keep these ACLs identical.

You should think of the IPv6 filter in the same way we think of filter in the hardware: if there are conflicts or the filter is wrongly defined, it's a bug.

cjdoucette replied:
Sorry, I should have updated this comment last night, but it ultimately proved unnecessary to use any kind of category. Like you said, the filter should operate like a hardware filter: any filters applied to the interface should be used for all blocks using IPv6 filters on that interface. That's why I was confused during our meeting -- I dropped the idea but forgot about this comment. :)

AltraMayor commented on gk/main.c:
```diff
@@ -634,6 +708,14 @@ gk_proc(void *arg)
+			if (ret < 0)
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: can't classify IPv6 packets\n");
+		}
```
1. In order to make this block even less intrusive for the blocks adding support for the IPv6 filter, put the whole block into an inline function. Being inline is important to avoid the cost of calling the new function and return if `acl.num` is not greater than zero.

2. `process_ipv6_acl()` should absorb the `RTE_LOG()` into its body.

```

AltraMayor commented on gk/main.c:
```diff
@@ -512,6 +586,9 @@ gk_proc(void *arg)
 		struct gk_cmd_entry *gk_cmds[GK_CMD_BURST_SIZE];
+		struct acl_search acl = {
+			.num = 0,
+		};
```
Create a macro for this definition and initialization. Motivation: being the less intrusive on the blocks that help the IPv6 filter.

cjdoucette commented:
I just rebased with the suggestions from this morning. You'll notice that I also updated the LACP patch, but the only change is to make the bonding modes in the configuration file to be round robin instead of LACP by default. This is because the interfaces will not send packets until LACP is setup, and making sure there's an LACP partner for every test we're running is a hassle, so it's easier just to keep it as round robin. The rest of the patch didn't change.

AltraMayor commented on gk/main.c:
```diff
@@ -534,13 +609,10 @@ gk_proc(void *arg)
-				if (submit_nd(pkt, &gk_conf->net->front) == -1)
-					drop_packet(pkt);
+			} else if (packet.flow.proto == ETHER_TYPE_IPv6 &&
+				     ipv6_if_configured(&gk_conf->net->front)) {
```
I think we should drop the second test. The functional blocks helping the IPv6 filter to work should make the minimum assumption on what's going on in the filter.

cjdoucette replied:
OK, I see the reasoning. But we need this test at some point, so I will push it to `process_pkts_ipv6_acl()` -- out of the view of the helping blocks.

AltraMayor replied:
It's okay to push this to `process_pkts_ipv6_acl()`, but I was thinking that if IPv6 is not configured, the ACLs will be empty, so they will free any packet that is pushed into classification. That is, the test if IPv6 is configured is redundant; it may give a small speedup at best.

AltraMayor commented on gk/main.c:
```diff
@@ -653,6 +727,9 @@ cleanup_gk(struct gk_config *gk_conf)
 	int i;
 
+	if (ipv6_if_configured(&gk_conf->net->front))
+		destroy_ipv6_acls(&gk_conf->net->front);
```
GK didn't include any ACL in the IPv6 filter, so why should it clear anything here? The same goes for the GT block.

AltraMayor commented on gk/main.c:
```diff
@@ -745,6 +822,22 @@ gk_stage2(void *arg)
+		int i;
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			unsigned int lcore = gk_conf->lcores[i];
+			ret = register_ipv6_acl(&gk_conf->net->front,
```
The LLS block should register the ACLs it needs, not the helping blocks. The same goes for the GT block.

cjdoucette replied:
The motivation to have the GK and GT blocks register is that they know which lcores, and therefore which sockets, need to have the ACL. What I'm seeing now is that the LLS block can be the only one that registers the ND rule, even if it installs it for sockets that will not use it. Is that correct?

AltraMayor replied:
Yes, that's correct. The problem you're pointed out is the same that we have with memory pools for mbufs. We have memory pools allocated for all sockets. You can an XXX comment saying that whenever the problem of memory pools per socket is solved, that solution solves the same issue for ACLs.

AltraMayor commented on gk/main.c:
```diff
@@ -745,6 +822,22 @@ gk_stage2(void *arg)
+				goto cleanup;
+			}
+		}
+		build_ipv6_acls(&gk_conf->net->front);
```
If all blocks that need ACL (i.e. LLS and CPS) register their ACL in stage1, `net_conf` can build the final ACL in stage2.  The same goes for the GT block.

cjdoucette replied:
I tried doing it that way, but there is a tricky sequence that needs to be observed. Calculating all the IPv6 addresses requires knowing the Ethernet address of the NIC, which can only be retrieved after the NIC is started (in stage 2). So the blocks that need the ACL are forced to register in stage 2. It would be nice to be able to build the ACLs just once in stage 3, but stage 3 is for launching the lcores of the blocks -- in other words, there's no stage 3 for the network configuration. So I think we could either change stage 3 to allow the ACL to be built once, or just allow the ACL to be re-built if needed when new registrations happen in stage 2.

AltraMayor replied:
We can have blocks registering ACLs at stage 2. All we need is to add the ACL building as the last step of stage 2, what we can accomplish adding a stage 2 call in `main/main.c:main()` just before calling `launch_gatekeeper()`.

cjdoucette replied:
Ah, even better. Thanks!

AltraMayor replied:
Don't forget to add a good comment explaining the reasoning in `main/main.c:main()`.

AltraMayor commented on gt/main.c:
```diff
@@ -505,30 +506,12 @@ gt_proc(void *arg)
-						    &gt_conf->net->front) == -1)
-						rte_pktmbuf_free(m);
+				if (pkt_info.outer_ip_ver == ETHER_TYPE_IPv6 &&
+				     ipv6_if_configured(&gt_conf->net->front)) {
```
We should drop the second test as of the argument in previous comment.

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,292 @@
+RTE_ACL_RULE_DEF(ipv6_rule, RTE_DIM(ipv6_defs));
+
+static void
+fill_nd_rule(struct ipv6_rule *rule, struct in6_addr *addr, int nd_type)
```
This function belongs to LLS.

cjdoucette commented:
I just rebased with the suggestions above and with Qiaobin's new patches.

AltraMayor commented on lib/net.c:
```diff
@@ -906,12 +917,10 @@ init_iface(struct gatekeeper_if *iface)
 
 	/* Initialize bonded port, if needed. */
-	if (iface->num_ports == 1)
+	if (iface->num_ports == 1 && iface->bonding_mode != BONDING_MODE_8023AD)
```
The exact negation of `iface->num_ports > 1 || iface->bonding_mode == BONDING_MODE_8023AD` is `iface->num_ports <= 1 || iface->bonding_mode == BONDING_MODE_8023AD`. Could you edit all occurrencies?

cjdoucette replied:
I see your point of getting the negation exactly right, but did you mean to write that it should be: `iface->num_ports <= 1 && iface->bonding_mode != BONDING_MODE_8023AD`?

AltraMayor replied:
Yes, your expression is the right one. Thank you.

AltraMayor commented on include/gatekeeper_acl.h:
```diff
@@ -0,0 +1,99 @@
+	GATEKEEPER_IPV6_NO_MATCH,
+	GATEKEEPER_IPV6_ACL_ND,
+	GATEKEEPER_IPV6_ACL_MAX,
+};
```
Why are these constants need? Why can't we have callbacks that are registered dynamically?

cjdoucette replied:
I don't think it has to be. If it's dynamic, then the blocks that want to install rules (like LLS) need to pass the rule generation function, the number of rules it will be generating, and the ACL callback function -- that's enough to install the rules and then call a function when one of those rules are matched. I'll add this in the next rebase.

AltraMayor commented on include/gatekeeper_acl.h:
```diff
@@ -0,0 +1,99 @@
+{
+	if (acl->num == 0 || !ipv6_if_configured(iface))
+		return;
+	process_ipv6_acl(iface, lcore, acl);
```
If `acl->num > 0` and `ipv6_if_configured(iface) == false`, packets must be freed.

cjdoucette commented:
OK, the suggestions have been integrated.

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,224 @@
+
+	/* Transmit separate buffers to registered ACL functions. */
+	for (i = 0; i < GATEKEEPER_IPV6_ACL_MAX; i++) {
+		ret = ((acl_cb_func)iface->acl_funcs[i])(pkts[i],
```
I think the cast `(acl_cb_func)` is not needed.

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	/* Transmit separate buffers to registered ACL functions. */
+	for (i = 0; i < GATEKEEPER_IPV6_ACL_MAX; i++) {
```
You can replace `GATEKEEPER_IPV6_ACL_MAX` with `iface->acl_count`.

AltraMayor commented on lls/main.c:
```diff
@@ -200,13 +206,14 @@ put_nd(struct in6_addr *ip_be, unsigned int lcore_id)
-submit_nd(struct rte_mbuf *pkt, struct gatekeeper_if *iface)
+submit_nd(struct rte_mbuf **pkts, int num_pkts, struct gatekeeper_if *iface)
 {
 	if (nd_enabled(&lls_conf)) {
```
Shouldn't this test be for ND at `iface` instead of ND in general? Even deeper, is this test needed at all since the ACLs are only installed if ND is enable at that interface?

cjdoucette replied:
Yes, I think you're right that it's unnecessary -- it's just un-caught artifact of the previous way of handling ND packets. Thanks!

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,224 @@
+	/* Transmit separate buffers to registered ACL functions. */
+	for (i = 0; i < GATEKEEPER_IPV6_ACL_MAX; i++) {
+		ret = ((acl_cb_func)iface->acl_funcs[i])(pkts[i],
+			num_pkts[i], iface);
```
`submit_nd()` does a lot of nothing when `num_pkts[i] == 0`, and this behavior is kind of expected to happen in other callback functions, so adding a test `num_pkts[i] > 0` before calling the callback function is a good idea.

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -197,6 +198,13 @@ struct gatekeeper_if {
+	/* Callback functions for each ACL rule type. */
+	acl_cb_func        acl_funcs[GATEKEEPER_IPV6_ACL_MAX];
+	/* Number of ACL types installed in @acl_funcs. */
+	unsigned int       acl_count;
```
The name `acl_count` is ambiguous. Suggestion: `acl_func_count`.

AltraMayor commented on lls/main.c:
```diff
@@ -500,12 +531,12 @@ lls_stage2(void *arg)
-	 * the DPDK libraries.
-	 */
-	if (lls_conf->nd_cache.iface_enabled(net_conf, &net_conf->back)) {
+	ret = register_nd_acl_rules(&net_conf->front);
```
1. The comment before this change, which I can't comment directly, should be updated since we don't need to know from where the ND packets are coming from with the IPv6 filter. We should think of it as hardware filters.

2. Before installing ARP filters, the code test if ARP is enabled at that interface, why should ND do the same?

AltraMayor commented on lls/main.c:
```diff
@@ -500,12 +531,12 @@ lls_stage2(void *arg)
+		return ret;
+
+	/* TODO Have a different block set up RSS on the back interface. */
+	if (ipv6_if_configured(&net_conf->back)) {
```
It's better to do like the ARP test above.

AltraMayor commented on lls/main.c:
```diff
@@ -500,12 +531,12 @@ lls_stage2(void *arg)
+	if (ret < 0)
+		return ret;
+
+	/* TODO Have a different block set up RSS on the back interface. */
```
`finalize_stage2()` could test if the back interface is enabled and no block set RSS for the back interface. For this case, `finalize_stage2()` could set RSS for LLS to take care of it.

cjdoucette replied:
I'm not sure I see the motivation for this. Should `finalize_stage2()` make sure the RSS on the front is set as well? It seems like there's no reason to assume the LLS wouldn't set it if the back interface is enabled, and the RT part of the GK block will fulfill this role a couple of patches down the line anyway.

AltraMayor replied:
Ok, let's wait to find the need for it.

I was thinking that there could be a setup in which no block sets RSS at the back. And as you pointed, this could eventually happen to the front as well.

AltraMayor commented on lls/main.c:
```diff
@@ -430,6 +399,68 @@ lls_proc(void *arg)
+
+	for (i = DST1_FIELD_IPV6; i <= DST4_FIELD_IPV6; i++) {
+		rule->field[i].value.u32 = rte_be_to_cpu_32(*ptr32);
+		rule->field[i].mask_range.u32 = 32;
```
Should the `32` be `0xFFFFFFFF`?

cjdoucette replied:
No, it's a configuration quirk. If the type of the field is `RTE_ACL_FIELD_TYPE_MASK`, then it represents a mask length.

AltraMayor replied:
I see the intention of avoiding a bitmask instead of a network mask. Thanks.

cjdoucette commented on lib/acl.c:
```diff
@@ -0,0 +1,224 @@
+
+	for (i = 0; i < numa_nodes; i++) {
+		int ret = rte_acl_add_rules(iface->ipv6_acls[i],
+			(struct rte_acl_rule *)ipv6_rules, num_rules);
```
I'm not sure why the ACL library does it this way, but `struct ipv6_acl_rule` is actually defined using a macro in include/gatekeeper_acl.h:

`RTE_ACL_RULE_DEF(ipv6_acl_rule, RTE_DIM(ipv6_defs));`

The `struct rte_acl_rule` is defined the same way (ctags won't find it either), but has no fields.

AltraMayor replied:
I got it. Thank you.

cjdoucette commented on lib/acl.c:
```diff
@@ -0,0 +1,224 @@
+{
+	struct rte_mbuf *pkts
+		[GATEKEEPER_IPV6_ACL_MAX][GATEKEEPER_MAX_PKT_BURST];
+	int num_pkts[GATEKEEPER_IPV6_ACL_MAX] = { 0 };
```
I think this kind of initialization is supported ([C99 6.7.8.21](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf#page=139), but I could be wrong about whether this situation is applicable to that.

cjdoucette replied:
In any case, I'm changing the size of the array to be `iface->acl_func_count`, and you can't initialize a dynamically-sized array, so I will be using `memset()`.

AltraMayor replied:
Although you ended needing to use the `memset()`, I do like that you supported your view citing the standard.

cjdoucette commented:
OK, all suggestions have been integrated, except where noted above.

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,243 @@
+
+	if (!ipv6_if_configured(iface)) {
+		drop_ipv6_acl_pkts(acl->mbufs, acl->num, iface);
+		return 0;
```
Whenever `process_ipv6_acl()` returns, it should zero `acl->num`. The reason for this is that `struct acl_search` may be defined outside of a loop. The only reason this problem has shown up is because `struct acl_search` is being defined inside loops.

AltraMayor commented on lib/acl.c:
```diff
@@ -0,0 +1,243 @@
+
+		ret = iface->acl_funcs[i](pkts[i], num_pkts[i], iface);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
```
This log entry should be a warning and there should be a way to know which packets need to be released. This way we can release the packets not already released/handled, and continue on with the other acl functions.

cjdoucette commented:
In the latest rebase, I made sure `acl->num = 0` before returning from `process_ipv6_acl()`, added `unlikely`s, and made it the responsibility of the ACL functions to free unhandled packets on error so that the rest of the ACL can be processed and Gatekeeper can proceed.

mengxiang0811 commented:
When I compiled the codebase with latest commits, and found an compiling error in line https://github.com/AltraMayor/gatekeeper/blob/master/lib/acl.c#L165. I think it's a typo, and changing ``WARN`` to ``WARNING`` works.

AltraMayor commented:
Hi @mengxiang0811,

Could you add a patch to fix that in your current pull request?

mengxiang0811 commented:
Hi @AltraMayor ,

Sure! Will do this.
